[Spring]Added support for composed annotations (@GetMapping,@PostMapping,@PutMapping,@DeleteMapping).

dRbAndrade opened the pull request:
I have noticed that when using the Spring generator, the API interface file would contain @RequestMapping(method = RequestMethod.<METHOD>) annotation. This would isn't exactly an issue but would be a code smell under sonar coding rule java:s4488 https://sonarsource.github.io/rspec/#/rspec/S4488. I tried to search for an existing way to change it but could not find it. So I went ahead and added the support myself.

<!-- Please check the completed items below -->
### PR checklist
 
- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [ ] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [ ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

martin-mfg commented on modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:
```diff
@@ -226,8 +226,15 @@ public interface {{classname}} {
     {{/swagger1AnnotationLibrary}}
     {{/implicitHeadersParams.0}}
+    {{#useApiComposedAnnotation}}
+    {{#lambda.composedannotationlambda}}{{httpMethod}}{{/lambda.composedannotationlambda}}
```
PS: This should also include updating the JavaDoc in ComposedRequestMappingAnnotationLambda.java.

dRbAndrade replied:
Agreed and done!

martin-mfg commented on modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:
```diff
@@ -226,8 +226,15 @@ public interface {{classname}} {
+    {{/useApiComposedAnnotation}}
+    {{^useApiComposedAnnotation}}
     @RequestMapping(
+    {{/useApiComposedAnnotation}}
+        {{^useApiComposedAnnotation}}
```
It looks like you can remove these 2 lines.

dRbAndrade replied:
Yeap, overlooked that haha

martin-mfg reviewed (changes requested):
Hi @dRbAndrade, thanks for the PR!

Could you please take care of the review comments below and also add your new lambda to this section in the docs:
https://github.com/OpenAPITools/openapi-generator/blob/4cffd32f87224e54a60813e1fc9da73788977e54/docs/templating.md?plain=1#L817

dRbAndrade commented:
> Hi @dRbAndrade, thanks for the PR!
> 
> Could you please take care of the review comments below and also add your new lambda to this section in the docs:
> 
> https://github.com/OpenAPITools/openapi-generator/blob/4cffd32f87224e54a60813e1fc9da73788977e54/docs/templating.md?plain=1#L817

Hello, @martin-mfg !
Sorry for the delay on the reply. Been busy.

I have updated with the requested changes, but I would advise against addting my lambda to the templating.md file because I can see all other lambdas are general purpose and can pretty much be used everywhere. Mine is very specific and has really no other use othe than this and should not be used by anyone else.

dRbAndrade commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:
```diff
@@ -98,6 +98,8 @@ public class SpringCodegen extends AbstractJavaCodegen
     public static final String OPTIONAL_ACCEPT_NULLABLE = "optionalAcceptNullable";
     public static final String USE_SPRING_BUILT_IN_VALIDATION = "useSpringBuiltInValidation";
+    public static final String USE_API_COMPOSED_ANNOTATION = "useApiComposedAnnotation";
+    public static final String API_COMPOSED_ANNOTATION_LAMBDA = "apiComposedAnnotationLambda";
```
Yeap, was trying something and left this out. Removed!

dRbAndrade commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:
```diff
@@ -224,6 +228,9 @@ public SpringCodegen() {
                 useSpringBuiltInValidation));
+        cliOptions.add(CliOption.newBoolean(USE_API_COMPOSED_ANNOTATION,
+                "Use @<method>Mapping instead of @RequestMapping(method = RequestMethod.<method>) when generating the API interfaces.",
+                useApiComposedAnnotation));
```
I agree this should be default behaviour. Since you're on board, changed the default to true

dRbAndrade commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/templating/mustache/ComposedRequestMappingAnnotationLambda.java:
```diff
@@ -0,0 +1,47 @@
+    @Override
+    public void execute(Template.Fragment fragment, Writer writer) throws IOException {
+            String text = fragment.execute();
+            writer.write(String.format("@%sMapping(",text.substring(0,1).toUpperCase()+text.substring(1).toLowerCase()));
```
Done as well

martin-mfg reviewed (changes requested):
> I have updated with the requested changes, but I would advise against addting my lambda to the templating.md file because I can see all other lambdas are general purpose and can pretty much be used everywhere. Mine is very specific and has really no other use othe than this and should not be used by anyone else.

Good point, I agree. Since the new lambda is not general purpose, I guess it should not be created in DefaultCodegen but instead in SpringCodegen. Like [this](https://github.com/OpenAPITools/openapi-generator/blob/2cff152e2a09d5c9297de4575c0ee552e4913651/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaMicronautAbstractCodegen.java#L700-L704).

Also, please run `./bin/utils/export_docs_generators.sh` to update the documentation.

dRbAndrade commented:
> > I have updated with the requested changes, but I would advise against addting my lambda to the templating.md file because I can see all other lambdas are general purpose and can pretty much be used everywhere. Mine is very specific and has really no other use othe than this and should not be used by anyone else.
> 
> Good point, I agree. Since the new lambda is not general purpose, I guess it should not be created in DefaultCodegen but instead in SpringCodegen. Like [this](https://github.com/OpenAPITools/openapi-generator/blob/2cff152e2a09d5c9297de4575c0ee552e4913651/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaMicronautAbstractCodegen.java#L700-L704).
> 
> Also, please run `./bin/utils/export_docs_generators.sh` to update the documentation.

Hey! @martin-mfg ! How you've been? So I updated the code and moved the lambda to the SpringCodegen class and ran the shell. But it generated LOTS of differences due to us setting the default behavior to true and I was not sure what to do so I just added them to the commit too. I think this was expected, I guess.

martin-mfg reviewed (approved):
> it generated LOTS of differences due to us setting the default behavior to true and I was not sure what to do so I just added them to the commit too. I think this was expected, I guess.

Yes, these expected changes need to be part of the PR. So thanks for adding them.

The PR looks good to me! Let's give wing328 some time to merge it.
